### PR TITLE
raise hand feature - react components

### PR DESCRIPTION
**Pull request by rzdor**
What

This is POC for Raise Hand feature.
I've added implementation for Raise/Lower Hand button to components.
Updated composites and Calling test app to use new element.
Added states and subscribers to hand updates for new published/removed states.

image
image

Why
Implementation for SDK raise hand feature

# How Tested

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-components/src/localization/locales/ar-SA/strings.json [hunk 1]
```diff
@@ -7,7 +7,9 @@
     "mutedIconLabel": "تم كتم الصوت",
     "displayNamePlaceholder": "مشارك غير مسمى",
     "participantStateRinging": "يتم الآن الاتصال...",
-    "participantStateHold": "قيد الانتظار"
+    "participantStateHold": "قيد الانتظار",
```
Comment (alkwa-msft) [inline packages/react-components/src/localization/locales/ar-SA/strings.json:10]: did we pull in these strings from TDBuild? We have a process for this. I would prefer to pull in TDBuild strings as a separate check in.

File: packages/react-components/src/components/RaiseHandButton.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    />
+  );
+};
+
+/* @conditional-compile-remove(raise-hand) */
+const raiseHandButtonStyles = (theme: Theme): IButtonStyles => ({
+  rootChecked: {
+    background: theme.palette.themePrimary,
+    color: DefaultPalette.white,
+    ':focus::after': { outlineColor: `${DefaultPalette.white} !important` } // added !important to avoid override by FluentUI button styles
```
Comment (alkwa-msft) [inline packages/react-components/src/components/RaiseHandButton.tsx:94]: Do not use !important. It is very confusing to build on top of. If we are already using it then please bring it up in the Feature Weekly Discussion.

**Review by alkwa-msft — APPROVED** [review]

**Comment by JamesBurnside** [comment]
Can you add screenshots of the changes? Are there any figma's to cross reference the design with?

File: packages/react-components/src/theming/icons.tsx [hunk 3]
```diff
@@ -235,6 +237,10 @@ export const DEFAULT_COMPONENT_ICONS = {
   ControlButtonParticipantsContextualMenuItem: <People20Regular />,
   ControlButtonScreenShareStart: <ShareScreenStart20Filled />,
   ControlButtonScreenShareStop: <ShareScreenStop20Filled />,
+  /* @conditional-compile-remove(raise-hand) */
+  ControlButtonRaiseHand: <HandLeft16Regular />,
```
Comment (dmceachernmsft) [inline packages/react-components/src/theming/icons.tsx:241]: We usually use the size 20 icons for the control bar is there a reason we went with the 16 here?

File: packages/react-components/src/types/ParticipantListParticipant.ts [hunk 4]
```diff
@@ -22,6 +22,19 @@ export type CallParticipantListParticipant = ParticipantListParticipant & {
...
+};
+
+/* @conditional-compile-remove(raise-hand) */
+/**
+ * Raised hand state with order
+ *
+ * @public
+ */
+export type RaisedHand = {
+  order: number;
```
Comment (dmceachernmsft) [inline packages/react-components/src/types/ParticipantListParticipant.ts:37]: nit: maybe call this something like `raisedHandOrderPosition`

File: packages/react-components/src/components/ParticipantList.tsx [hunk 5]
```diff
@@ -218,6 +243,20 @@ export const ParticipantList = (props: ParticipantListProps): JSX.Element => {
         });
       }
 
+      /* @conditional-compile-remove(raise-hand) */
+      const callingParticipant = participant as CallParticipantListParticipant;
+      /* @conditional-compile-remove(raise-hand) */
+      if (callingParticipant.raisedHand && onLowerHands) {
```
[the remark this answers is not in the document]
Comment (rzdor) [reply packages/react-components/src/components/ParticipantList.tsx:249]: That is correct anyone on the call can remote state for other users

File: packages/react-components/src/components/ParticipantList.tsx [hunk 6]
```diff
@@ -218,6 +243,20 @@ export const ParticipantList = (props: ParticipantListProps): JSX.Element => {
         });
       }
 
+      /* @conditional-compile-remove(raise-hand) */
+      const callingParticipant = participant as CallParticipantListParticipant;
```
[the remark this answers is not in the document]
Comment (rzdor) [reply packages/react-components/src/components/ParticipantList.tsx:247]: renamed

File: packages/react-components/src/types/ParticipantListParticipant.ts (hunk 4 above)
Comment (rzdor) [reply packages/react-components/src/types/ParticipantListParticipant.ts:37]: changed to 20

File: packages/react-components/src/components/ParticipantItem.tsx [hunk 7]
```diff
@@ -76,6 +76,12 @@ export interface ParticipantItemStrings {
   /* @conditional-compile-remove(PSTN-calls) */
   /** String shown when `participantState` is `Hold` */
   participantStateHold?: string;
+  /* @conditional-compile-remove(raise-hand) */
+  /** Label for the remove button in participant menu  */
+  lowerParticipantHandButtonLabel: string;
```
[the remark this answers is not in the document]
Comment (rzdor) [reply packages/react-components/src/components/ParticipantItem.tsx:81]: I will added them, currently no toast for published states

File: packages/react-components/src/components/VideoTile.tsx [hunk 8]
```diff
@@ -341,11 +349,11 @@ export const VideoTile = (props: VideoTileProps): JSX.Element => {
           background: theme.palette.neutralLighter,
           borderRadius: theme.effects.roundedCorner4
         },
-        isSpeaking && {
+        (isSpeaking || /* @conditional-compile-remove(raise-hand) */ raisedHand) && {
           '&::after': {
             content: `''`,
             position: 'absolute',
-            border: `0.25rem solid ${theme.palette.themePrimary}`,
+            border: `0.25rem solid ${isSpeaking ? theme.palette.themePrimary : theme.palette.orangeLighter}`,
```
Comment (JamesBurnside) [inline packages/react-components/src/components/VideoTile.tsx:356]: this color looks incorrect in your test app

| testapp | figma |
| -- | -- |
| image | image |

There likely isn't an existing theme slot for this color so we will need to add a new semantic slot to the theme provider similar to how we have done for the end call button. I think we should add something like: `raiseHandGold` to this 
image

File: packages/react-components/src/components/ParticipantList.tsx [hunk 9]
```diff
@@ -147,6 +153,22 @@ const onRenderParticipantDefault = (
...
               <Icon iconName="ParticipantItemMicOff" className={iconStyles} ariaLabel={strings.mutedIconLabel} />
             )}
+            {
+              /* @conditional-compile-remove(raise-hand) */ callingParticipant.raisedHand && (
+                <Stack horizontal={true} tokens={{ childrenGap: '0.2rem' }}>
+                  <Stack.Item>
+                    <Text>{callingParticipant.raisedHand?.raisedHandOrderPosition}</Text>
+                  </Stack.Item>
+                  <Stack.Item>
+                    <Icon
```
Comment (JamesBurnside) [inline packages/react-components/src/components/ParticipantList.tsx:163]: Color for this looks incorrect, should be the golden color:

| figma | testapp |
| -- | -- |
| image | image |

File: packages/react-components/src/components/ParticipantItem.tsx (hunk 7 above)
Comment (JamesBurnside) [inline packages/react-components/src/components/ParticipantItem.tsx:81]: I see interface changes in `ParticipantItem.tsx` but no functional change in `ParticipantItem.tsx`. If someone is just using the participantitem component and not the participantlist componet, how would they use these? Some of the logic in ParticipantList may need to be moved to ParticipantItem (or vice-versa, these strings in participantItem need moved to participantsList)

**Review by JamesBurnside — CHANGES_REQUESTED** [review]

File: change/@azure-communication-react-c6978365-5dd8-44c8-b2b4-85c8f248f726.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "update components",
```
[the remark this answers is not in the document]
Comment (rzdor) [reply change/@azure-communication-react-c6978365-5dd8-44c8-b2b4-85c8f248f726.json]: updated

File: packages/calling-component-bindings/src/participantListSelector.ts [hunk 11]
```diff
@@ -15,7 +15,7 @@ import { getRole } from './baseSelectors';
 import { CallParticipantListParticipant } from '@internal/react-components';
 import { _isRingingPSTNParticipant, _updateUserDisplayNames } from './utils/callUtils';
 import { memoizedConvertAllremoteParticipants } from './utils/participantListSelectorUtils';
-/* @conditional-compile-remove(rooms) */
+/* @conditional-compile-remove(raise-hand) */
```
Comment (JamesBurnside) [inline packages/calling-component-bindings/src/participantListSelector.ts:18]: If its used for both features, and both to the conditional compilation line
```suggestion
/* @conditional-compile-remove(rooms) */ /* @conditional-compile-remove(raise-hand) */
```

Comment (JamesBurnside) [reply packages/calling-component-bindings/src/participantListSelector.ts:18]: (for this line and throughout)

File: packages/communication-react/review/stable/communication-react.api.md [hunk 12]
```diff
@@ -512,6 +512,7 @@ export interface CallingTheme {
         callRedDark: string;
         callRedDarker: string;
         iconWhite: string;
+        raiseHandGold: string;
```
Comment (JamesBurnside) [inline packages/communication-react/review/stable/communication-react.api.md:515]: this should be conditionally compiled out of stable

File: packages/react-components/src/components/RaiseHandButton.test.tsx [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+/* @conditional-compile-remove(raise-hand) */
+import { RaiseHandButton } from './RaiseHandButton';
+/* @conditional-compile-remove(raise-hand) */
+import { createTestLocale, renderWithLocalization } from './utils/testUtils';
+/* @conditional-compile-remove(raise-hand) */
+import { registerIcons } from '@fluentui/react';
+/* @conditional-compile-remove(raise-hand) */
+import { screen } from '@testing-library/react';
+
+describe('RaiseHandButton strings should be localizable and overridable', () => {
```
Comment (JamesBurnside) [inline packages/react-components/src/components/RaiseHandButton.test.tsx:15]: 💖

File: packages/react-components/src/components/RaiseHandButton.tsx [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    />
+  );
+};
+
+/* @conditional-compile-remove(raise-hand) */
+const raiseHandButtonStyles = (theme: Theme): IButtonStyles => ({
+  rootChecked: {
+    background: theme.palette.themePrimary,
+    color: DefaultPalette.white,
+    ':focus::after': { outlineColor: `${DefaultPalette.white}` } // added !important to avoid override by FluentUI button styles
```
Comment (JamesBurnside) [inline packages/react-components/src/components/RaiseHandButton.tsx:94]: remove no longer relevant comment
```suggestion
    ':focus::after': { outlineColor: `${DefaultPalette.white}` }
```

Comment (JamesBurnside) [reply packages/react-components/src/components/RaiseHandButton.tsx:94]: (same with below)

File: packages/react-components/src/components/assets/RaisedHandIcon.tsx [hunk 15]
```diff
@@ -0,0 +1,835 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (JamesBurnside) [inline packages/react-components/src/components/assets/RaisedHandIcon.tsx:1]: Hmm this is a pretty big svg.. gzip size of this is 5KB cc @PorterNan if this is OK or should we find an alternative solution

File: packages/communication-react/package.json [hunk 16]
```diff
@@ -64,8 +64,9 @@
     "copy-preprocess": "cpx \"../*/preprocessed/**\" ./preprocess-dist && renamer --find preprocessed --replace src ./preprocess-dist/**/*",
     "build": "rushx check-deps && rushx _by-flavor \"rushx _build:by-flavor\"",
     "build:cjs": "rollup -c --silent",
-    "build:esm": "rushx _if-preprocess && rushx copy-original && rushx preprocess && rushx copy-preprocess  && ttsc -project tsconfig.preprocess.json || (if-env COMMUNICATION_REACT_FLAVOR=beta && ttsc) && rushx copy-northstar",
+    "build:esm": "rushx _if-preprocess && rushx copy-original && rushx preprocess && rushx copy-preprocess  && ttsc -project tsconfig.preprocess.json || (if-env COMMUNICATION_REACT_FLAVOR=beta && ttsc) && rushx copy-northstar && copy-react-components",
```
Comment (JamesBurnside) [inline packages/communication-react/package.json:67]: revert this change now we moved to svg as react component?

**Review by JamesBurnside — APPROVED** [review]
Open question on whether we should use fluent icon vs custom icon. But we can update in a subsequent PR. This text also looks a little out of place now  (I think someone being called shouldn't have any mic icon)
image

But also can be done outside of this PR as this PR is already v big

File: packages/react-components/src/components/assets/RaisedHandIcon.tsx (hunk 15 above)
Comment (PorterNan) [reply packages/react-components/src/components/assets/RaisedHandIcon.tsx:1]: Found this one to compress the svg a bit, from 47k to 13k - https://www.iloveimg.com/compress-image/compress-svg
Not sure how it works, but if that could make the size smaller then let's do it

13k is still a bit too huge for one icon, but at least a lot better than 50k

**Pull request merged**